Document the new 9.1 OpenTelemetry meters (#6719)
Particular/docs.particular.net: nservicebus/operations/opentelemetry.md

summary: Observability of NServiceBus endpoints with OpenTelemetry
component: core
reviewed: 2022-06-29
versions: '[8,)'
related:
 - samples/open-telemetry
---

NServiceBus version 8 and above supports [OpenTelemetry](https://opentelemetry.io/docs/instrumentation/net/) through traces, metrics, and logging.

> Enable OpenTelemetry instrumentation in NServiceBus:

Reviewer: Why was this inlined? I just split this up further in another PR that conflicts with this change, so we'll have to coordinate on which structure to keep. Given what I've added now in the tracing section, I think it's getting large enough to benefit from splitting up. See https://github.com/Particular/docs.particular.net/pull/6718

Author: Laila, go ahead with that PR. Once that's merged, I'll rebase (or maybe even recreate) this one to align with those larger changes. No worries.

snippet: opentelemetry-enablemeters

partial: meters

See the [OpenTelemetry samples](/samples/open-telemetry/) for instructions on how to send metric information to different tools.

## Logging

NServiceBus supports logging out of the box. To collect OpenTelemetry-compatible logging in NServiceBus endpoints, it's possible to configure the endpoint to connect traces and logging when using `Microsoft.Extensions.Logging` package. See the [_Connecting OpenTelemetry traces and logs_ sample](/samples/open-telemetry/logging) for more details.

## Alignment of host identifier

It is recommended to align the instance identifier between NServiceBus and OpenTelemetry so all logs, metrics, traces and audit messages can be correlated by a host (instance) if needed.

> [!NOTE]
> The OpenTelemetry specification recommends this to be a random uuid. However, it may also be a [deterministic uuid v5](https://opentelemetry.io/docs/specs/semconv/attributes-registry/service/#service-attributes) (i.e. hash of machine name and  endpointname).

> NServiceBus adds an [host identifier to all audit messages](/nservicebus/hosting/override-hostid.md) and this instance identifier is also used to show [performance metrics for each running instance in ServicePulse](/monitoring/metrics/in-servicepulse.md). The [instance identifier used for ServicePulse value can be overriden](/monitoring/metrics/install-plugin.md#configuration-instance-id).

Reviewer 2: ```suggestion
NServiceBus adds a [host identifier to all audit messages](/nservicebus/hosting/override-hostid.md) and this instance identifier is also used to show [performance metrics for each running instance in ServicePulse](/monitoring/metrics/in-servicepulse.md). The [instance identifier used for ServicePulse value can be overriden](/monitoring/metrics/install-plugin.md#configuration-instance-id).
```

> OpenTelemetry also allows to customize the instance id used for `service.instance.id` in various ways.

Reviewer 2: ```suggestion
OpenTelemetry also allows customizing the instance id used for `service.instance.id` in various ways.
```

Consider aligning the instance ID used by OpenTelemetry and ServiceControl metrics API.

#### Example

> ```c#

Reviewer 2: Make legit snippet? (Not sure if this would be an 8 and 9 snippet or a 9.1 only snippet)

// Generate (deterministic) instance ID shared by all components

// Generate deterministic uuid v4 viahttps://github.com/Faithlife/FaithlifeUtility/blob/master/src/Faithlife.Utility/GuidUtility.cs
var deterministicValue = "MyEndpoint@" + Dns.GetHostName();
Guid serviceInstanceId = GuidUtility.Create(new Guid("4d63009a-8d0f-11ee-aad7-4c796ed8e320", deterministicValue)) // or Guid.NewGuid()

// OpenTelemetry
services.AddOpenTelemetry()

>     .ConfigureResource(rb => rb.AddService("MyService", serviceInstanceId: serviceInstanceId.ToString()))

Reviewer 2: Guessing `rb` means resource builder? `builder` might be better without being overlong?
```suggestion
    .ConfigureResource(builder => builder.AddService("MyService", serviceInstanceId: serviceInstanceId.ToString()))
```
